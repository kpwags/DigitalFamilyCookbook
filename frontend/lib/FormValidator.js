import validator from 'validator';

class FormValidator {
    static validateEmail(value) {
        if (typeof value !== 'string' || !validator.isEmail(value)) {
            return false;
        }

        return true;
    }

    static validateInvitationCode(value) {
        if (!this.validateNotEmpty(value)) {
            return {
                valid: false,
                message: 'Invitation code is required'
            };
        }

        const regEx = new RegExp('^[A-Za-z0-9_]+$');
        if (!regEx.test(value)) {
            return {
                valid: false,
                message: 'Invitation codes can only contain letters, numbers, and underscores'
            };
        }

        if (value.trim().length < 1 || value.trim().length > 20) {
            return {
                valid: false,
                message: 'Invitation codes must be between 1 and 20 characters'
            };
        }

        return {
            valid: true,
            message: ''
        };
    }

    static validateNotEmpty(value) {
        if (typeof value !== 'string' || value.trim().length === 0) {
            return false;
        }

        return true;
    }

    static validateSetLength(value, fieldName = 'value', minLength = 1, maxLength = 25) {
        if (!this.validateNotEmpty(value)) {
            return {
                valid: false,
                message: `${fieldName} is required`
            };
        }

        if (value.trim().length < minLength || value.trim().length > maxLength) {
            return {
                valid: false,
                message: `${fieldName} must be between ${minLength} and ${maxLength} characters`
            };
        }

        return {
            valid: true,
            message: ''
        };
    }

    static validateUsername(value) {
        if (!this.validateNotEmpty(value)) {
            return {
                valid: false,
                message: 'Username is required'
            };
        }

        const regEx = new RegExp('^[A-Za-z0-9_]+$');
        if (!regEx.test(value)) {
            return {
                valid: false,
                message: 'Username can only contain letters, numbers, and underscores'
            };
        }

        if (value.trim().length < 1 || value.trim().length > 20) {
            return {
                valid: false,
                message: 'Username must be between 1 and 20 characters'
            };
        }

        return {
            valid: true,
            message: ''
        };
    }

    static validatePassword(value1, value2) {
        if (typeof value1 !== 'string' || typeof value2 !== 'string' || (value1 === '' && value2 === '')) {
            return {
                valid: false,
                message: 'Password is required'
            };
        }

        if (value1.trim() !== value2.trim()) {
            return {
                valid: false,
                message: 'Passwords do not match'
            };
        }

        if (value1.trim().length < 8) {
            return {
                valid: false,
                message: 'Passwords must be at least 8 characters long'
            };
        }

        return {
            valid: true,
            message: ''
        };
    }
}

export { FormValidator };
