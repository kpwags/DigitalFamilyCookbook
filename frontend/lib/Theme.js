const Theme = {
    black: 'hsl(0, 0%, 22.4%)',
    green: 'hsl(140, 100%, 25%)',
    lightGreen: 'hsl(136, 44%, 47%)',
    paleGreen: 'hsl(136, 100%, 88%)',
    darkGreen: 'hsl(140, 100%, 15%)',
    bs: '0 12px 24px 0 hsla(0, 0%, 0%, 0.2)'
};

export { Theme };
