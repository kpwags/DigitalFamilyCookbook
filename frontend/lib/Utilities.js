import Router from 'next/router';

class Utilities {
    static setTableRowActive(tableId, rowId) {
        const table = document.getElementById(tableId);
        const tbody = table.getElementsByTagName('tbody')[0];

        // TODO: Find better way?
        // eslint-disable-next-line no-restricted-syntax
        for (const row of tbody.getElementsByTagName('tr')) {
            row.className = '';
        }

        document.getElementById(rowId).className = 'selected';
    }

    // TODO: Add Message
    static invalidateField(elementId, errorMessage = '') {
        document.getElementById(elementId).className = 'errored';
        document.getElementById(elementId).parentElement.className = 'errored';

        if (errorMessage !== '') {
            document.getElementById(`${elementId}-message`).innerText = errorMessage;
            document.getElementById(`${elementId}-message`).style.display = 'block';
        }
    }

    static invalidateFieldByClass(className) {
        const elements = document.getElementsByClassName(className);

        for (let i = 0; i < elements.length; i += 1) {
            elements[i].className = `${className}  errored`;
        }
    }

    static resetFieldByClass(className) {
        const elements = document.getElementsByClassName(className);

        for (let i = 0; i < elements.length; i += 1) {
            let elementClassName = elements[i].className;
            elementClassName = elementClassName.replace('errored', '');
            elements[i].className = elementClassName;
        }
    }

    static resetField(elementId) {
        let { className } = document.getElementById(elementId);
        className = className.replace('errored', '');
        document.getElementById(elementId).className = className;

        let parentClassName = document.getElementById(elementId).parentElement.className;
        parentClassName = parentClassName.replace('errored', '');
        document.getElementById(elementId).parentElement.className = parentClassName;
        document.getElementById(`${elementId}-message`).style.display = 'none';
        document.getElementById(`${elementId}-message`).innerText = '';
    }

    static getNextAvailableValue(arr, idx = 'key') {
        let nextValue = 0;
        arr.forEach(item => {
            if (item[idx] >= nextValue) {
                nextValue = item[idx];
            }
        });

        nextValue += 1;

        return nextValue;
    }

    static convertCookingTime(time) {
        let timeHours = 0;
        let timeMinutes = 0;

        let returnString = '';

        if (time !== null && parseInt(time, 10) > 0) {
            timeHours = Math.floor(time / 60);
            timeMinutes = time % 60;

            if (timeHours === 1) {
                returnString += `${timeHours} hour`;
            } else if (timeHours > 1) {
                returnString += `${timeHours} hours`;
            }

            if (timeHours > 0 && timeMinutes > 0) {
                returnString += ', ';
            }

            if (timeMinutes > 0) {
                returnString += `${timeMinutes} minutes`;
            }
        }

        return returnString;
    }

    static convertIDArray(arr) {
        const newArr = [];
        arr.forEach(i => {
            newArr.push(i.id);
        });

        return newArr;
    }

    static toTitleCase(str) {
        return str
            .toLowerCase()
            .split(' ')
            .map(word => {
                return word.replace(word[0], word[0].toUpperCase());
            })
            .join(' ');
    }

    static goToRecipe(id) {
        Router.push({
            pathname: '/recipe',
            query: { id }
        });
    }

    static toggleElement(elementId) {
        if (document.getElementById(elementId).style.display !== 'none') {
            document.getElementById(elementId).style.display = 'none';
        } else {
            document.getElementById(elementId).style.display = 'block';
        }
    }

    static toggleMobileMenu() {
        document.getElementById('hamburgermenu').classList.toggle('change');

        if (document.getElementById('mobilemenu').style.left === '0px') {
            document.getElementById('mobilemenu').style.left = '-250px';
        } else {
            document.getElementById('mobilemenu').style.left = '0px';
        }
    }

    static toggleHeaderMenu(menuId) {
        if (
            document.getElementById(menuId).style.display === 'none' ||
            document.getElementById(menuId).style.display === ''
        ) {
            document.getElementById(menuId).style.display = 'block';
        } else {
            document.getElementById(menuId).style.display = 'none';
        }
    }

    static hideHeaderMenu(menuId) {
        document.getElementById(menuId).style.display = 'none';
    }

    static formatDate(date) {
        const formattedDate = new Date(date);
        return formattedDate.toLocaleDateString('en-US');
    }

    static boxColors = ['gray', 'blue', 'green', 'orange', 'purple'];
}

export { Utilities };
