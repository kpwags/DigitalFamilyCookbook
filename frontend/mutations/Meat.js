import gql from 'graphql-tag';

const CREATE_MEAT_MUTATION = gql`
    mutation CREATE_MEAT_MUTATION($name: String!) {
        createMeat(name: $name) {
            id
            name
        }
    }
`;

const UPDATE_MEAT_MUTATION = gql`
    mutation UPDATE_MEAT_MUTATION($id: ID!, $name: String) {
        updateMeat(id: $id, name: $name) {
            id
            name
        }
    }
`;

const DELETE_MEAT_MUTATION = gql`
    mutation DELETE_MEAT_MUTATION($id: ID!) {
        deleteMeat(id: $id) {
            id
        }
    }
`;

export { CREATE_MEAT_MUTATION };
export { UPDATE_MEAT_MUTATION };
export { DELETE_MEAT_MUTATION };
