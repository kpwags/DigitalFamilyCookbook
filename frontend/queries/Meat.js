import gql from 'graphql-tag';

const ALL_MEATS_QUERY = gql`
    query {
        meats {
            id
            name
        }
    }
`;

export { ALL_MEATS_QUERY };
