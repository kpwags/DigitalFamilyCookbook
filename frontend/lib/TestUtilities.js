/* eslint-disable import/no-extraneous-dependencies */
import casual from 'casual';
import PropTypes from 'prop-types';
import { MockedProvider } from '@apollo/react-testing';
import { ThemeProvider } from 'styled-components';
import { Theme } from './Theme';

const MockedThemeProvider = (props) => {
    const { mocks = null, addTypename = false, children } = props;
    return (
        <ThemeProvider theme={Theme}>
            <MockedProvider mocks={mocks} addTypename={addTypename}>
                {children}
            </MockedProvider>
        </ThemeProvider>
    );
};

MockedThemeProvider.propTypes = {
    children: PropTypes.node,
    mocks: PropTypes.array,
    addTypename: PropTypes.bool,
};

const TestAdmin = () => ({
    __typename: 'User',
    id: casual.uuid,
    name: casual.name,
    username: casual.username,
    email: casual.email,
    bio: 'I am just a test admin',
    image: 'user.jpg',
    largeImage: 'user-lg.jpg',
    permissions: ['USER', 'ADMIN'],
});

const TestCategory = () => ({
    __typename: 'Category',
    id: casual.uuid,
    name: casual.word,
    createdAt: casual.date,
});

const TestDirection = () => ({
    __typename: 'Direction',
    id: casual.uuid,
    name: casual.words(),
    sortOrder: casual.integer,
});

const TestIngredient = () => ({
    __typename: 'Ingredient',
    id: casual.uuid,
    name: casual.words(),
    sortOrder: casual.integer,
});

const TestInvitationCode = () => ({
    __typename: 'InvitationCode',
    id: casual.uuid,
    code: casual.word,
    createdAt: casual.date,
});

const TestMeat = () => ({
    __typename: 'Meat',
    id: casual.uuid,
    name: casual.words(),
    createdAt: casual.date,
});

const TestUser = () => ({
    __typename: 'User',
    id: casual.uuid,
    name: casual.name,
    username: casual.word,
    email: casual.email,
    bio: 'I am just a test user',
    image: 'user.jpg',
    largeImage: 'user-lg.jpg',
    permissions: ['USER'],
});

const TestRecipe = () => ({
    __typename: 'Recipe',
    id: casual.uuid,
    name: casual.name,
    public: casual.boolean,
    source: casual.words(),
    sourceUrl: casual.url,
    time: casual.integer,
    activeTime: casual.integer,
    servings: casual.integer,
    calories: casual.integer,
    carbohydrates: casual.integer,
    fat: casual.integer,
    sugar: casual.integer,
    cholesterol: casual.integer,
    fiber: casual.integer,
    image: 'recipe.jpg',
    largeImage: 'recipe-large.jpg',
    ingredients: [TestIngredient(), TestIngredient()],
    directions: [TestDirection(), TestDirection()],
    meats: [TestMeat(), TestMeat()],
    categories: [TestCategory(), TestCategory()],
    user: TestUser(),
});

const TestRecipesList = () => [TestRecipe(), TestRecipe(), TestRecipe(), TestRecipe(), TestRecipe()];

export { MockedThemeProvider, TestAdmin, TestCategory, TestDirection, TestIngredient, TestInvitationCode, TestMeat, TestRecipe, TestRecipesList, TestUser };
