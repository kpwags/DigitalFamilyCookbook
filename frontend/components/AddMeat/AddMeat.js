import React, { useState } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { toast } from 'react-toastify';
import PropTypes from 'prop-types';
import { Form } from '../Form/Form';
import { TextInput } from '../TextInput/TextInput';
import { ErrorMessage } from '../ErrorMessage/ErrorMessage';
import { CREATE_MEAT_MUTATION } from '../../mutations/Meat';
import { ALL_MEATS_QUERY } from '../../queries/Meat';
import { FormValidator } from '../../lib/FormValidator';

const AddMeat = (props) => {
    const [name, setName] = useState('');
    const [error, setError] = useState(null);
    const [nameError, setNameError] = useState('');

    const [createMeat, { loading: createLoading, error: createError }] = useMutation(CREATE_MEAT_MUTATION, {
        refetchQueries: [{ query: ALL_MEATS_QUERY }],
        onCompleted: (data) => {
            if (error) {
                if (props.onError) {
                    props.onError(error);
                }
            } else {
                setName('');
                setNameError('');

                toast(`${data.createMeat.name} added successfully`);

                if (props.onComplete) {
                    props.onComplete();
                }
            }
        },
        onError: (err) => {
            if (props.onError) {
                props.onError(err);
            } else {
                setError(err);
            }
        },
    });

    const validateForm = () => {
        let isValid = true;

        if (!FormValidator.validateNotEmpty(name)) {
            setNameError('Name is required');
            isValid = false;
        }

        return isValid;
    };

    const cancelAdd = (e) => {
        e.preventDefault();
        setName('');
        setNameError('');

        if (props.onCancel) {
            props.onCancel();
        }
    };

    return (
        <Form
            data-test="form"
            id="create-meat-form"
            onSubmit={async (e) => {
                e.preventDefault();

                setError(null);

                if (validateForm()) {
                    await createMeat({ variables: { name } }).catch((err) => {
                        if (props.onError) {
                            props.onError(err);
                        } else {
                            setError(err);
                        }
                    });
                }
            }}
        >
            <ErrorMessage error={error || createError} />
            <fieldset disabled={createLoading} aria-busy={createLoading}>
                <TextInput
                    name="name"
                    label="Name"
                    id="add-meat-name"
                    validationRule="notempty"
                    value={name}
                    error={nameError}
                    onChange={(e) => {
                        setName(e.target.value);
                    }}
                />

                <button type="submit">Sav{createLoading ? 'ing' : 'e'}</button>
                <button type="button" onClick={cancelAdd}>
                    Cancel
                </button>
            </fieldset>
        </Form>
    );
};

AddMeat.propTypes = {
    onComplete: PropTypes.func,
    onCancel: PropTypes.func,
    onError: PropTypes.func,
};

export { AddMeat };
