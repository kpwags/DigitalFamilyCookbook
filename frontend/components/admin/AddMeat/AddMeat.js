import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import { Form } from '../../styles/Form';
import { ErrorMessage } from '../../elements/ErrorMessage';
import { CREATE_MEAT_MUTATION } from '../../../mutations/Meat';
import { ALL_MEATS_QUERY } from '../../../queries/Meat';
import { FormValidator } from '../../../lib/FormValidator';
import { Utilities } from '../../../lib/Utilities';

class AddMeat extends Component {
    state = {
        name: '',
        error: null
    };

    handleChange = e => {
        const { id, name, type, value } = e.target;
        const val = type === 'number' ? parseFloat(value) : value;
        this.setState({ [name]: val }, () => {
            this.validate(id, val);
        });
    };

    validate = (fieldId, value) => {
        // eslint-disable-next-line default-case
        switch (fieldId) {
        case 'add-meat-name':
            if (!FormValidator.validateNotEmpty(value)) {
                Utilities.invalidateField('add-meat-name', 'Name is required.');
            } else {
                Utilities.resetField('add-meat-name');
            }
            break;
        }
    };

    validateForm = () => {
        let isValid = true;

        let { name } = this.state;
        if (name === '') {
            name = document.getElementById('add-meat-name').value;
        }

        if (!FormValidator.validateNotEmpty(name)) {
            Utilities.invalidateField('add-meat-name', 'Name is required.');
            isValid = false;
        }

        return isValid;
    };

    hideAddForm = () => {
        document.getElementById('create-meat-header-form').style.display = 'none';
        this.setState({ name: '' });
        document.getElementById('create-meat-form').reset();
    };

    cancelAddMeat = e => {
        e.preventDefault();
        this.hideAddForm();
    };

    render() {
        return (
            <Mutation
                mutation={CREATE_MEAT_MUTATION}
                variables={this.state}
                refetchQueries={[{ query: ALL_MEATS_QUERY }]}
            >
                {(createMeat, { loading, error }) => (
                    <Form
                        data-test="form"
                        method="POST"
                        id="create-meat-form"
                        onSubmit={async e => {
                            e.preventDefault();

                            this.setState({ error: null });

                            if (this.validateForm()) {
                                await createMeat().catch(err => {
                                    this.setState({ error: err });
                                });

                                if (this.state.error === null) {
                                    this.hideAddForm();
                                }
                            }
                        }}
                    >
                        <ErrorMessage error={error || this.state.error} />
                        <fieldset disabled={loading} aria-busy={loading}>
                            <label htmlFor="name">
                                Name
                                <input
                                    type="text"
                                    id="add-meat-name"
                                    name="name"
                                    placeholder="Name"
                                    required
                                    value={this.state.name}
                                    onChange={this.handleChange}
                                    onBlur={e => {
                                        e.preventDefault();
                                        this.validate('add-meat-name', this.state.name);
                                    }}
                                />
                                <div className="error-text" id="add-meat-name-message" />
                            </label>
                            <button type="submit">Sav{loading ? 'ing' : 'e'}</button>
                            <button type="button" onClick={this.cancelAddMeat}>
                                Cancel
                            </button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }
}

export { AddMeat };
