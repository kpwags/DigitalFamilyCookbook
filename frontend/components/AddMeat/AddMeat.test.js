import { render, fireEvent, act } from '@testing-library/react';
import { ApolloProvider } from '@apollo/react-hooks';
import { createMockClient } from 'mock-apollo-client';
import { CREATE_MEAT_MUTATION } from '../../mutations/Meat';
import { ALL_MEATS_QUERY } from '../../queries/Meat';
import { TestMeat, MockedThemeProvider } from '../../lib/TestUtilities';
import { AddMeat } from './AddMeat';

describe('<AddMeat/>', () => {
    const meat = TestMeat();

    const mockClient = createMockClient();

    // Mock the result of the login mutation
    const createMeatMutationHandler = jest.fn().mockResolvedValue({
        data: {
            createMeat: {
                id: meat.id,
                name: meat.name,
            },
        },
    });

    const allMeatsQueryHandler = jest.fn().mockResolvedValue({
        data: {
            meats: [TestMeat(), TestMeat(), TestMeat()],
        },
    });

    mockClient.setRequestHandler(CREATE_MEAT_MUTATION, createMeatMutationHandler);
    mockClient.setRequestHandler(ALL_MEATS_QUERY, allMeatsQueryHandler);

    test('it renders the input', async () => {
        const { findByLabelText } = render(
            <MockedThemeProvider>
                <AddMeat />
            </MockedThemeProvider>
        );

        await findByLabelText(/Name/);
    });

    test('it creates a meat when the form is submited', async () => {
        await act(async () => {
            const { getByText, getByLabelText } = render(
                <ApolloProvider client={mockClient}>
                    <AddMeat />
                </ApolloProvider>
            );

            fireEvent.change(await getByLabelText(/Name/), {
                target: {
                    value: meat.name,
                },
            });

            fireEvent.click(await getByText(/Save/));
        });

        expect(createMeatMutationHandler).toBeCalledWith({
            name: meat.name,
        });

        expect(allMeatsQueryHandler).toBeCalledTimes(1);
    });

    test('it alerts the user a meat name is required if left blank', async () => {
        const { findByText, getByLabelText } = render(
            <ApolloProvider client={mockClient}>
                <AddMeat />
            </ApolloProvider>
        );

        fireEvent.blur(await getByLabelText(/Name/));

        // Assert that the error message was displayed
        await findByText(/Name is required/);
    });
});
