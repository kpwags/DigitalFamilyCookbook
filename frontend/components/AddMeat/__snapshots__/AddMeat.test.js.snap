// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMeat/> should match snapshot 1`] = `
<form
  className="Form-sc-1jw7vbt-0 jSygZL"
  data-test="form"
  id="create-meat-form"
  method="POST"
  onSubmit={[Function]}
>
  <ErrorMessage
    error={null}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="name"
    >
      Name
      <input
        id="add-meat-name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Name"
        required={true}
        type="text"
        value=""
      />
      <div
        className="error-text"
        id="add-meat-name-message"
      />
    </label>
    <button
      type="submit"
    >
      Sav
      e
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </fieldset>
</form>
`;
